Tables guideline n'at (#488)
brightlayer-ui/doc-it: src/docs/patterns/tables.mdx

        TableCellStatus,
        TableCellIcon,
        TableCellMultiSelect,
        TableCellStripe,
    ]}
    captionsUnderImages={[
        'Text content. The alternating background color is optional.',
        'Numerical value. Note that both the header and the text cells are right-aligned.',
        'Text fields. Note that text fields are right-aligned as they are expecting numerical values. These rows are also taller than standard table rows.',
        'Switches and checkboxes.',
        'Icons and tags.',
        'Icon buttons at the end of the table row for touch screens.',
        `Checkboxes to select rows.`,
        `Status stripes for row items with special statuses.`,
    ]}
    gridImageProps={{ sm: 6, md: 4 }}
/>

You should seek to make your tables both easy to navigate and effective to use.

<ImageGrid
    images={[TableCellUnderlinedField, TableCellImage]}
    captionsUnderImages={[
        'CAUTION: Be careful when using underlined text fields in a table, as their understrokes compete with the line strokes between each table row.',
        `AVOID: Images can heighten table rows. Consider using a list instead.`,
    ]}
/>

# Behaviors

## Row Actions

> Similar to lists, tables also have row actions. Common row actions include edit, delete and view details. They are usually placed at the end of a table row.

Reviewer: ```suggestion
Similar to lists, tables can also have row-level actions. Common actions include edit, delete, and view details. They are usually placed at the end of a table row.
```

<ImageGrid
    images={[ActionAtRowEnd, ActionOnHover]}
    captionsUnderImages={[
        `Row actions can be placed at the end of the row.`,

>         `DO: On desktop, you can choose to have row actions visible only when the mouse is hovering on the row. The buttons won't compete for rooms with the rest of the table.`,

Reviewer: ```suggestion
        `DO: On desktop, you can choose to have row actions visible only when the mouse is hovering on the row. This will save space and make your tables cleaner and easier to read.`,
```

    ]}
/>

<ImageGrid
    images={[Drag1, Drag2]}
    caption={[
        `Drag-and-drop actions are usually placed at the beginning of a row. If you have row numbers, you may choose to display the drag handle on hover.`,
    ]}
/>

## Batch Actions

> If your application allows users to perform batch actions to their data set, such as deleting 15 items at once, you can add a checkbox column on the left rim of the table.

Reviewer: ```suggestion
If your application allows users to perform batch actions on their data set, such as deleting 15 items at once, you can add a checkbox column on the left side of the table for handling selection.
```

        `Buttons become enabled when rows are checked. Optionally, use theme color to highlight the selected row, and label the button with the number of items selected to account for pagination.`,
    ]}
/>

<!--

Note: I think this is a good suggestion, but might be too high of a requirement for our adopters

You may change the clickable area when users are cognitively loaded and busy deciding which row to check. When they enter this "batch action mode", increase the row selection area so that it is easier to select.

<ImageGrid
    images={[ClickArea1, ClickArea2]}
    captionsUnderImages={[
        `When nothing is selected, the selection target area is limited to the region around the checkbox.`,
        `When a row is selected, the application enters this "batch action mode", and users can select other rows by clicking anywhere on those rows.`,
    ]}
/>
-->

## Sticky Header Row

When a table gets long, your users should not need to scroll back to the top to view the table header.

<ImageGrid
    images={[StickyHeader]}
    caption={[`Use sticky headers to help your users navigate through the table.`]}
    regularWidth
/>

> ## Expanded Row

Reviewer: This one worries me a bit from an implementation standpoint. It feels less like a table with this kind of behavior. How would this affect sorting and filtering?

Author: The expanded part would be viewed as "attached" to the original content, and would not affect the sorting / filtering at all.

> Occasionally, table rows may expand to allow users _take a quick peek_ into a small amount of row details.

Reviewer: ```suggestion
Occasionally, table rows may expand to allow users to take a quick peek into additional row details.
```

<ImageGrid
    images={[ExpandedRow]}

>     caption={[`Users can quickly glance into a small amount of row details.`]}

Reviewer: ```suggestion
    caption={[`Users can quickly glance into a small amount of additional row data.`]}
```

    regularWidth
/>

<ImageGrid
    images={[ExpandedRowTooComplicated, ExpandedRowActions]}
    captionsUnderImages={[
        `DONT: Do not include complicated contents in an expanded row, as this will break the efficiency of a table. In this case, you should design a dedicated page for the content.`,
        `DONT: Do not surprise users by hiding high priority actions in the expanded row space.`,
    ]}
/>

## Horizontal Scrolling

> When your table overflows horizontally, you should allow your users to scroll horizontally to view more about the table. However, this interaction is less ergonomically convenient when users are not using a touch screen or a trackpad. See the [Responsive](#responsive) section below.

Reviewer: I think we should be discouraging horizontal scrolling and encouraging people to only show as much data as is necessary.

Author: I added a `CAUTION:` tag to the section

    caption={[`Place horizontal scroll bar between rows and the footer / pagination area.`]}
    regularWidth
/>

<ImageGrid
    images={[StickyCheckbox, StickyRowAction]}
    captionsUnderImages={[
        `The checkbox column should be sticky to the left as the user scrolls to the right.`,
        `The action column, if designed to be always visible, should be sticky to the right. Do not include too many row actions that cover excessive table body area.`,
    ]}
/>

## Searching and Filtering

There are many different ways to perform searching and filtering with table data. Below are a few examples:

<ImageGrid
    images={[Filter1, Filter2, Filter3, Filter4]}
    captionsUnderImages={[
        `A simple search field embedded in a table.`,
        `A search field outside a table.`,
        `Filter criteria listed as chips.`,
        `A "saved filter" dropdown.`,
    ]}
/>

<ImageGrid
    images={[AdvancedTableFilter]}
    caption={[
        `You may combine multiple interactions (searching and filtering) and input methods (keyboard navigations, keyboard typing, mouse clicking) together for an "advanced search".`,
    ]}
    regularWidth
/>

## Transition Animation

> Often, table data look highly similar between rows and can disorient your users. You are encouraged to use animations in your applications to help users understand certain transitions.

Reviewer: ```suggestion
Table rows often look very similar and can be difficult for users to differentiate. You should make use of animations and transitions to help users understand where certain actions are taking place.
```

Author: Since you changed it from "we encourage" to "you should", I am going to add `DO:` to the image captions to emphasize that this is a must.

<ImageGrid
    images={[AdditionAnimation, SubstractionAnimation]}
    captionsUnderImages={[
        `Transition animation for adding a row to the table. Consider using a background color to temporarily highlight the newly-added row.`,
        `Transition animation for removing a row from the table.`,
    ]}
/>

<Divider />

# Tables vs. Lists

<ImageGrid
    images={[ListVTableTable, ListVTableList]}
    caption={['A same set of data may be presented in tables (#1) or lists (#2), depends on the use cases.']}
/>

> In most cases, we encourage you to organize your data into lists instead of tables. However, there is no clearly defined boundary between lists and tables. Below are a few factors you should consider when deciding how you want to present your data.

Reviewer: ```suggestion
In most cases, we encourage you to organize your data into lists instead of tables. However, the distinction between the two is not black and white. Below are a few factors you should consider when deciding how to present your data.
```

## Arrangement

> Tables expose each data point into their own columns, allowing users to easily look up a row starting from any column. Lists, on the other hand, group their content row-by-row.

Reviewer: ```suggestion
Tables place more emphasis on columns, allowing users to easily find a row by any of its associated data points. Lists, on the other hand, can be less structured and tend to treat each item as a cohesive whole rather than a collection of individual data values.
```

<ImageGrid
    images={[ArrangeTable, ArrangeList]}
    captionsUnderImages={[

>         'Tables take up more space and is often the main focus of the page they are on.',

Reviewer: ```suggestion
        'Tables usually take up more horizontal space and are often the main focus of the page they are on.',
```

        'Lists are more horizontally compact and are frequently used for dashboards. Typography changes can give the text more visual hierarchy.',
    ]}
/>

> Because of this, tables tend to take up a lot of horizontal space, therefore they are almost always the only main content on the page of a desktop application. You should not use tables in dashboards, where many widgets are competing for the space; you should also avoid using tables on mobile applications, where the viewport size is limited.

Reviewer: ```suggestion
Because of this, tables tend to take up much more horizontal space. Therefore, they are almost always the primary content on the page where they reside. You should not use tables in dashboards, where many widgets are competing for the space; you should also avoid using tables on mobile applications, where the viewport size is limited.
```

<ImageGrid
    images={[DashboardTable, PageBodyTable]}
    captionsUnderImages={[
        'AVOID: Avoid using tables in dashboards. Use lists instead.',
        'DO: Tables are usually the only main content on a page.',
    ]}
/>

> A "tabular list" or a structured list that attempts to combine tables and lists together might result in a design that does not pick up the advantages of either pattern.

Reviewer: I might just leave this bit out entirely if it's not something we encourage

Author: I've seen several teams doing it; would be good to give our official opinion on this.

<ImageGrid
    images={[TabularList]}
    captionsUnderImages={[
        'CAUTION: Be careful when combining tables and lists together, as this diminishes the value of both.',
    ]}
    gridContainerProps={{ style: { justifyContent: 'flex-start' } }}
/>

## Eye Movement

> Tables facilitate users who want to work across multiple different properties of a data point. Users often navigate down a column, find the desired value, jump to another column, navigating further down, before they exit their path. This is called a [lawn mower pattern](https://www.nngroup.com/articles/lawn-mower-pattern/). On the other hand, lists often have a primary property — can be icons, avatars, or primary list texts — that the user would rely on for navigation, a scanning behavior known as the [layer cake pattern](https://www.nngroup.com/articles/layer-cake-pattern-scanning/).

Reviewer: ```suggestion
Tables facilitate users who want to work across multiple different properties of an item. Users often navigate down a column, find the desired value, jump to another column, navigating further down, before they exit their path. This is called a [lawn mower pattern](https://www.nngroup.com/articles/lawn-mower-pattern/). On the other hand, lists often have a primary property — these can be icons, avatars, or primary list text — that the user would rely on for navigation, a scanning behavior known as the [layer cake pattern](https://www.nngroup.com/articles/layer-cake-pattern-scanning/).
```

    caption={[
        `If you do an eye tracking between when the user starts looking at their data set and when they decide to perform some actions, you would notice users' eye sights roam differently in these two structures.`,
        '1: In a table, users can start from any column / property, and navigate horizontally or vertically to the targeted item (lawn mower pattern).',
        '2: In a list, users rely heavily on a primary property to navigate (layer cake pattern).',
    ]}
/>

If you expect your users to compare across multiple parameters before they can decide which row to interact with, you should use a table. If they mostly read your data row-by-row, you shall try structuring your data as a list.

<ImageGrid
    images={[ColumnOrder1, ColumnOrder2]}
    caption={[
        `Users are more likely to find a user by their name or their role, not by their email address or phone number. Therefore, you should arrange the name and the role columns to the left, followed by phone number and email (#1). Email addresses and phone numbers should not show up before the role column (#2).`,
    ]}
    captionsUnderImages={['', 'DONT:']}
/>

## Property Name

> In tables, you would call out properties of each data entry only once in the header row. In lists, however, either the values are clear enough on their own without explanations, or property names have to be included in every list item.

Reviewer: ```suggestion
In tables, each data point is labeled only once in the table header. In lists, however, the values are either clear enough on their own without explanations, or they have to be included in each list item.
```

<ImageGrid
    images={[ImplicitValueTable]}
    captionsUnderImages={[`The values are explained by the table header row.`]}
    gridContainerProps={{ style: { justifyContent: 'flex-start' } }}
/>
<ImageGrid
    images={[ImplicitValueList1, ImplicitValueList2]}
    captionsUnderImages={[
        `DONT: It is not clear what those values in the subtitle mean.`,
        `CAUTION: Be careful when calling out property names in lists, as they weaken the scanability.`,
    ]}
/>

## Consumer-Facing Applications

Avoid using tables in consumer-facing applications. Lists mentally match real-world individual objects, while tables can be more abstract. Use tables only if your users really need to examine technical details.

<Divider />

# Responsive

> In general, **we do not encourage tables on mobile platforms**, as table-related interactions tend to be complicated for small viewports, and mobile phones also often lack the computational power to process massive data set.

Reviewer: ```suggestion
In general, **we do not encourage tables on mobile platforms**, as table-related interactions tend to be complicated for small viewports, and mobile phones also often lack the computational power to process massive data sets.
```

<ImageGrid
    images={[TableOnMobile]}
    regularWidth
    caption={`Users can only see a very small portion of the table through mobile devices.`}
/>

## Understanding Users' Goal

<ImageGrid images={[UserInField]} regularWidth caption={`While on foot, it is very hard to examine small details.`} />

> Before you design a table structure for a mobile application, you should conduct user researches and have a clear understanding of users' intentions. When they land on this page, Do they desire to do complicated cross-comparison between multiple rows and columns? Is a full table necessary?

Reviewer: ```suggestion
Before you design a table structure for a mobile application, you should conduct user research and have a clear understanding of users' intentions. When they land on this page, do they need to do complicated cross-comparisons between multiple rows and columns? Is a full table necessary?
```

> For industrial applications, your users would often be using a mobile app when they are out in the field. They are paying more attention to their surroundings, so keep the mobile app interaction minimal.

Reviewer: This is not technically correct. There are a lot of industrial/monitoring applications that are not mobile. I would content that mobile apps are the new kid on the block in that regard.

Author: Meant to say that, for industrial mobile apps, when your users are using them, they are often on foot. 🙈

## Breaking Down Tables into Lists

If a user only needs a few parameters to identify the correct data entry and make decisions, a list with a few selected data points would suffice.

<ImageGrid
    images={[HomePagePreview, TableStackedAsList]}
    captionsUnderImages={[
        `On a crowded page, show a preview of the table content, with the option to view the full table in a dedicated page.`,
        `Stack each property into a list item, with the ability to click into a list item to view the full item details.`,
    ]}
/>

## Pagination

> Pagination is especially helpful on hand-held devices to reduce the load of both the server and the client-side render.

Reviewer: This is a little confusing...this is coming after the section where we just said you should not use tables on mobile, but now we are saying pagination is great for tables on mobile.

        `DONT: Do not surprise users by hiding high priority actions in the expanded row space.`,
    ]}
/>

## Horizontal Scrolling

We strongly encourage you to design your application to only show as much data as is necessary.

When your table overflows horizontally, you should allow your users to scroll horizontally to view more about the table. However, this interaction is less ergonomically convenient when users are not using a touch screen or a trackpad. See the [Responsive](#responsive) section below.

The application may display a scroll bar to suggest that there are more contents to the right.

<ImageGrid
    images={[HorizontalScrollBar]}
    caption={[
        `CAUTION: Be careful not to place too many columns to overflow the table. If a horizontal scroll is necessary, place a horizontal scroll bar between rows and the footer / pagination area.`,
    ]}
    regularWidth
/>

<ImageGrid
    images={[StickyCheckbox, StickyRowAction]}
    captionsUnderImages={[
        `The checkbox column should be sticky to the left as the user scrolls to the right.`,

>         `The action column, if designed to be always visible, should be sticky to the right. Do not include too many row actions that cover excessive table body area.`,

Reviewer: ```suggestion
        `The action column, if designed to be always visible, should be sticky to the right. Do not include too many actions or they will consume a significant portion of your available space.`,
```

<ImageGrid
    images={[Filter1, Filter2, Filter3, Filter4]}
    captionsUnderImages={[
        `A simple search field embedded in a table.`,
        `A search field outside a table.`,
        `Filter criteria listed as chips.`,
        `A "saved filter" dropdown.`,
    ]}
/>

<ImageGrid
    images={[AdvancedTableFilter]}
    caption={[
        `You may combine multiple interactions (searching and filtering) and input methods (keyboard navigations, keyboard typing, mouse clicking) together for an "advanced search".`,
    ]}
    regularWidth
/>

## Transition Animation

Table rows often look very similar and can be difficult for users to differentiate. You should make use of animations and transitions to help users understand where certain actions are taking place.

<ImageGrid
    images={[AdditionAnimation, SubstractionAnimation]}
    captionsUnderImages={[
        `DO: Transition animation for adding a row to the table. Consider using a background color to temporarily highlight the newly-added row.`,
        `DO: Transition animation for removing a row from the table.`,
    ]}
/>

<Divider />

# Tables vs. Lists

<ImageGrid
    images={[ListVTableTable, ListVTableList]}

>     caption={['A same set of data may be presented in tables (#1) or lists (#2), depends on the use cases.']}

Reviewer: ```suggestion
    caption={['The same set of data may be presented in tables (#1) or lists (#2), depends on the use case.']}
```

        'DO: Tables are usually the only main content on a page.',
    ]}
/>

A "tabular list" or a structured list that attempts to combine tables and lists together might result in a design that does not pick up the advantages of either pattern.

<ImageGrid
    images={[TabularList]}
    captionsUnderImages={[
        'CAUTION: Be careful when combining tables and lists together, as this diminishes the value of both.',
    ]}
    gridContainerProps={{ style: { justifyContent: 'flex-start' } }}
/>

## Eye Movement

Tables facilitate users who want to work across multiple different properties of an item. Users often navigate down a column, find the desired value, jump to another column, navigating further down, before they exit their path. This is called a [lawn mower pattern](https://www.nngroup.com/articles/lawn-mower-pattern/). On the other hand, lists often have a primary property — these can be icons, avatars, or primary list text — that the user would rely on for navigation, a scanning behavior known as the [layer cake pattern](https://www.nngroup.com/articles/layer-cake-pattern-scanning/).

<ImageGrid
    images={[ReadTable, ReadList]}
    caption={[

>         `If you do an eye tracking between when the user starts looking at their data set and when they decide to perform some actions, you would notice users' eye sights roam differently in these two structures.`,

Reviewer: ```suggestion
        `If you conduct eye tracking as a users attempts to perform some action, you would notice users scan differently depending on the data structure.`,
```

        '1: In a table, users can start from any column / property, and navigate horizontally or vertically to the targeted item (lawn mower pattern).',
        '2: In a list, users rely heavily on a primary property to navigate (layer cake pattern).',
    ]}
/>

> If you expect your users to compare across multiple parameters before they can decide which row to interact with, you should use a table. If they mostly read your data row-by-row, you shall try structuring your data as a list.

Reviewer: ```suggestion
If you expect your users to compare across multiple parameters before deciding which row to interact with, you should use a table. If they primarily read your data row-by-row, you should try structuring your data as a list.
```

<ImageGrid
    images={[ImplicitValueTable]}
    captionsUnderImages={[`The values are explained by the table header row.`]}
    gridContainerProps={{ style: { justifyContent: 'flex-start' } }}
/>
<ImageGrid
    images={[ImplicitValueList1, ImplicitValueList2]}
    captionsUnderImages={[
        `DONT: It is not clear what those values in the subtitle mean.`,
        `CAUTION: Be careful when calling out property names in lists, as they weaken the scanability.`,
    ]}
/>

## Consumer-Facing Applications

Avoid using tables in consumer-facing applications. Lists mentally match real-world individual objects, while tables can be more abstract. Use tables only if your users really need to examine technical details.

<Divider />

# Responsive

In general, **we do not encourage tables on mobile platforms**, as table-related interactions tend to be complicated for small viewports, and mobile phones also often lack the computational power to process massive data sets.

<ImageGrid
    images={[TableOnMobile]}
    regularWidth

>     caption={`Users can only see a very small portion of the table through mobile devices.`}

Reviewer: ```suggestion
    caption={`Users can only see a very small portion of the table on mobile devices.`}
```

/>

> ## Understanding Users' Goal

Reviewer: ```suggestion
## Understanding Users' Goals
```

<ImageGrid images={[UserInField]} regularWidth caption={`While on foot, it is very hard to examine small details.`} />

Before you design a table structure for a mobile application, you should conduct user research and have a clear understanding of users' intentions. When they land on this page, do they need to do complicated cross-comparisons between multiple rows and columns? Is a full table necessary?

> For an industrial mobile application, users would often be on foot and out in the field when they interact with the app. They are paying more attention to their surroundings, so keep the mobile app interaction minimal.

Reviewer: ```suggestion
For applications where users will be on foot and out in the field while they interact with the app, keep the interactions minimal so users can pay sufficient attention to their surroundings.
```

## Breaking Down Tables into Lists

If a user only needs a few parameters to identify the correct data entry and make decisions, a list with a few selected data points would suffice.

<ImageGrid
    images={[HomePagePreview, TableStackedAsList]}
    captionsUnderImages={[
        `On a crowded page, show a preview of the table content, with the option to view the full table in a dedicated page.`,
        `Stack each property into a list item, with the ability to click into a list item to view the full item details.`,
    ]}
/>

## Pagination

> If tables become necessary in your mobile application, pagination is especially helpful to reduce the load of both the server and the client-side render.

Reviewer: I think there is probably a stronger case for infinite scroll with virtualization (or progressive loading) on mobile versus pagination. Vertical scrolling is a very intuitive interaction on mobile. I don't think I would like having to hit the next button 9 times to get to the end of my list...I would rather fling the heck out of the screen and zoom to the bottom.

Author: I will add this back... I didn't add inf scroll bc Ian hated it.

<ImageGrid
    images={[MobilePagination]}
    captionsUnderImages={[
        `Pagination is often used to reduce the page load. If not conflicting with other gestures, consider also adding left / right swipes to go to the previous / next page.`,
    ]}
    gridContainerProps={{ style: { justifyContent: 'flex-start' } }}
/>

> Be careful using pagination with multi-selection, as it is unclear whether the selection gets preserved when the page number changes.

## Native Mobile Gestures

In native mobile applications, you may also consider embedding action gestures common for touch screens, such as long-press and swipe.

<MaterialDesignDescription title={'Gestures'} url={`https://material.io/design/interaction/gestures.html`} />

However, these gestures should never be the only way to perform an action. For example, if your application is able to delete a table row, users may be able to either (1) left-swipe to bring up a contextual menu with a delete option, or (2) click into the target row, which opens a new page that contains a delete button.

## Landscape Mode

> If a full table view is desired on mobile devices, consider offering a landscape view so that more columns would be visible at once.

Reviewer: ```suggestion
If a full table view is desired on mobile devices, consider offering a landscape view so that more columns are visible at once.
```

<ImageGrid images={[LandscapeTable]} caption={`Use the landscape mode for viewing the full table.`} regularWidth />

> However, most applications are meant to be used mainly on the portrait mode. In this case, you need to call out the option.

Reviewer: ```suggestion
However, many applications are meant to be used primarily in portrait mode. In this case, you need to call out the option.
```

<ImageGrid
    images={[InviteToLandscapeX, InviteToLandscapeCountdown]}
    caption={`For example, you can use a chip button to "invite" users to the landscape view.`}
/>

<Divider />

# Design Specifications

<ImageGrid
    images={[SpecSize, SpecTypography, SpecPagination]}
    captionsUnderImages={[`Dimensions`, `Colors and typography`, `Pagination`]}
    regularWidth
/>

<MaterialDesignDescription
    title={'Data Tables'}
    description={`For additional instructions, refer to Material Design's design specs.`}
    url={`https://material.io/components/data-tables#specs`}
/>

<Divider />

# Developers

Use the following components to implement this pattern:

**Angular** and **Ionic**:

-   Angular Material
    -   [Table](https://material.angular.io/components/table/overview)

**React**:

-   Material UI

>     -   [Table](https://mui.com/components/tables/)

Reviewer: Might also want to include a link to the XDataGrid: https://mui.com/components/data-grid/
